### PR TITLE
EPMEDU-617: Add Verification Screen

### PR DESCRIPTION
**Pull request by diantyn**
**Done:**
- Verification screen UI part
- Timer logic has been added

**Todo:**
- Strings will be added in next PR
- Amplify also will be added in next PR
- Unit tests will be added in next PR

### CODE REVIEW

File: UIComponents/Sources/UIComponents/Components/Inputs/VerificationInput/VerificationInputView.swift [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    // MARK: - Configuration
+    public func configure(_ model: Model) {
+        inputViews.removeAll()
+        containerView.arrangedSubviews.forEach { $0.removeFromSuperview() }
+        model.code.enumerated().forEach { digit in
```
Comment (igorTimofiejczyk) [inline UIComponents/Sources/UIComponents/Components/Inputs/VerificationInput/VerificationInputView.swift:207]: Do we need ```enumerated()``` here?

File: animeal/src/Modules/Verification/Model/VerificationModel.swift [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    // MARK: - Responses
+    var fetchNewCodeResponse: ((VerificationModelCode) -> Void)?
+    var fetchNewCodeResponseTimeLeft: ((VerificationModelTimeLeft) -> Void)?
+
+    // MARK: - Initialization
+    init() {
+        self.timeLeft = CurrentValueSubject<Int, Never>(30)
+        self.timerTimeout = 30
+        self.cancellables = []
+        self.expectedCode = "1111"
```
Comment (ihartsimafeichyk) [inline animeal/src/Modules/Verification/Model/VerificationModel.swift:32]: ```self.expectedCode = "1111"```  It's for testing needs?

File: animeal/src/Modules/Verification/VerificationViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                    true
+                )
+            }
+        }
+        model.fetchNewCodeResponseTimeLeft = { [weak self] modelTimeLeft in
+            guard let self = self else { return }
+            DispatchQueue.main.async {
+                self.onResendCodeHasBeenPrepared?(
+                    VereficationViewResendCode(
+                        title: "Resend code in",
```
Comment (ihartsimafeichyk) [inline animeal/src/Modules/Verification/VerificationViewModel.swift:52]: raw Strings

File: animeal/src/Modules/Verification/VerificationViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                )
+            }
+        }
+    }
+
+    func load() {
+        onHeaderHasBeenPrepared?(
+            VerificationViewHeader(
+                title: "Enter verification code",
+                subtitle: "Code was sent to: 558 49-99-69"
```
Comment (ihartsimafeichyk) [inline animeal/src/Modules/Verification/VerificationViewModel.swift:66]: I believe it's for testing needs?

File: animeal/src/Modules/Verification/VerificationViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+         formatter.allowedUnits = [.minute, .second]
+         formatter.unitsStyle = .positional
+         formatter.zeroFormattingBehavior = []
+         return formatter
+     }()
+
+    // MARK: - Dependencies
+    private let model: VerificationModelProtocol
+
+
```
Comment (EugeniyGubar) [inline animeal/src/Modules/Verification/VerificationViewModel.swift:16]: extra line

File: animeal/src/Modules/Verification/VerificationViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    // MARK: - State
+    var onHeaderHasBeenPrepared: ((VerificationViewHeader) -> Void)?
+    var onCodeHasBeenPrepared: ((VerificationViewCode, Bool) -> Void)?
+    var onResendCodeHasBeenPrepared: ((VereficationViewResendCode) -> Void)?
+
+
```
Comment (EugeniyGubar) [inline animeal/src/Modules/Verification/VerificationViewModel.swift:22]: extra line

File: animeal/src/Modules/Verification/VerificationViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    init(
+        model: VerificationModelProtocol
+    ) {
+        self.model = model
+        setup()
+    }
+
+    // MARK: - Life cycle
+    func setup() {
+        model.fetchNewCodeResponse = { [weak self] modelCode in
```
Comment (EugeniyGubar) [inline animeal/src/Modules/Verification/VerificationViewModel.swift:33]: I think, this logic could be moved to VerificationModuleAssembler.swift

Comment (diantyn) [reply animeal/src/Modules/Verification/VerificationViewModel.swift:33]: No, we could have some private logic here (ex. mapping)

File: UIComponents/Sources/UIComponents/Components/Inputs/VerificationInput/VerificationInputView.swift (hunk 1 above)
Comment (diantyn) [reply UIComponents/Sources/UIComponents/Components/Inputs/VerificationInput/VerificationInputView.swift:207]: Yes, we use an element and also offset(as index) here

File: animeal/src/Modules/Verification/VerificationViewModel.swift (hunk 7 above)
Comment (EugeniyGubar) [reply animeal/src/Modules/Verification/VerificationViewModel.swift:33]: How about this solution:

VerificationViewModel.swift
```swift
func fetchNewCodeResponse(code: VerificationModelCode) {
    DispatchQueue.main.async {
        self.onCodeHasBeenPrepared?(
            VerificationViewCode(
                state: VerificationViewCodeState.normal,
                items: code.items.map {
                    VerificationViewCodeItem(identifier: $0.identifier, text: $0.text)
                }
            ),
            true
        )
    }
}
```

VerificationModuleAssembler.swift
```swift
model.fetchNewCodeResponse = { [weak viewModel] modelCode in
    viewModel?.fetchNewCodeResponse(modelCode)
}
```

**Review by igorTimofiejczyk — APPROVED** [review]

**Review by igorTimofiejczyk — APPROVED** [review]

File: animeal/src/Modules/Verification/VerificationContract.swift [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    var fetchNewCodeResponseTimeLeft: ((VerificationModelTimeLeft) -> Void)? { get set }
+
+    func isValidationNeeded(_ code: VerificationModelCode) -> Bool
+    func validateCode(_ code: VerificationModelCode) -> Bool
+    func fetchInitialCode()
+    func fetchNewCode()
+}
+
+// MARK: - Assembler
+protocol VerificationAssembler {
```
Comment (EugeniyGubar) [inline animeal/src/Modules/Verification/VerificationContract.swift:44]: We already have this protocol for all assemblers
```swift
// MARK: - Assembling
protocol Assembling {
    func assemble() -> UIViewController
}
```

Comment (EugeniyGubar) [reply animeal/src/Modules/Verification/VerificationContract.swift:44]: Sorry, I haven't commited it yet:)

File: animeal/src/Modules/Verification/VerificationViewModel.swift (hunk 7 above)
Comment (diantyn) [reply animeal/src/Modules/Verification/VerificationViewModel.swift:33]: It's not necessary now to do such changes. We have only one action here, if an amount of actions raises, we will divide logic into several methods.

**Review by EugeniyGubar — APPROVED** [review]

**Pull request merged**